File: rust/integration-gateway/src/refresh/providers.rs
use crate::config::Config;

/// Resolved OAuth provider config for a refreshable integration kind.
pub struct Provider {
    pub token_url: String,
    pub client_id: String,
    pub client_secret: String,
}

/// Resolve the OAuth provider for `kind`, or `None` when the kind isn't a gateway-supported OAuth2
/// provider or its client credentials aren't configured.
///
/// Initial supported set is the generic `grant_type=refresh_token` providers (client id/secret in
/// the form body). Providers with non-standard refresh flows — Slack (DB-sourced creds), the
/// HTTP-Basic ones (reddit/pinterest/stripe), TikTok (different host + `client_key`), Bing (resends
/// scope), Jira (rotates refresh tokens), Meta (long-lived exchange), and the service-account /
/// GitHub kinds — stay on the Django beat for now.
///
/// `config.refresh_token_url_override`, when set, replaces the provider URL for every kind (local
/// e2e against a mock token endpoint).
pub fn provider_for(kind: &str, config: &Config) -> Option<Provider> {
    let (default_url, client_id, client_secret): (&str, &str, &str) = match kind {
        "hubspot" => (
            "https://api.hubapi.com/oauth/v1/token",
            config.hubspot_client_id.as_str(),
            config.hubspot_client_secret.as_str(),
        ),
        "salesforce" => (
            "https://login.salesforce.com/services/oauth2/token",
            config.salesforce_client_id.as_str(),
            config.salesforce_client_secret.as_str(),
        ),
        "google-ads" => (
            "https://oauth2.googleapis.com/token",
            config.google_ads_client_id.as_str(),
            config.google_ads_client_secret.as_str(),
        ),
        "google-analytics" => (
            "https://oauth2.googleapis.com/token",
            config.google_analytics_client_id.as_str(),
            config.google_analytics_client_secret.as_str(),
        ),
        "google-search-console" => (
            "https://oauth2.googleapis.com/token",
            config.google_search_console_client_id.as_str(),
            config.google_search_console_client_secret.as_str(),
        ),
        "google-sheets" => (
            "https://oauth2.googleapis.com/token",
            config.google_sheets_client_id.as_str(),
            config.google_sheets_client_secret.as_str(),
        ),
        _ => return None,
    };

    if client_id.is_empty() || client_secret.is_empty() {
        return None;
    }

    let token_url = if config.refresh_token_url_override.is_empty() {
        default_url.to_string()
    } else {
        config.refresh_token_url_override.clone()
    };

    Some(Provider {
        token_url,
        client_id: client_id.to_string(),
        client_secret: client_secret.to_string(),
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use envconfig::Envconfig;

    // A Config with all defaults, credential fields explicitly cleared so ambient env can't leak in.
    // Fields are pub, so tests set exactly what they exercise — no process-env mutation (race-free).
    fn base() -> Config {
        let mut c = Config::init_from_env().unwrap();
        c.hubspot_client_id = String::new();
        c.hubspot_client_secret = String::new();
        c.refresh_token_url_override = String::new();
        c
    }

    #[test]
    fn unknown_kind_has_no_provider() {
        let config = base();
        assert!(provider_for("slack", &config).is_none());
        assert!(provider_for("github", &config).is_none());
    }

    #[test]
    fn kind_without_credentials_is_skipped() {
        assert!(provider_for("hubspot", &base()).is_none());
    }

    #[test]
    fn configured_kind_resolves_and_override_applies() {
        let mut config = base();
        config.hubspot_client_id = "cid".to_string();
        config.hubspot_client_secret = "secret".to_string();
        config.refresh_token_url_override = "http://localhost:9999/token".to_string();

        let provider = provider_for("hubspot", &config).expect("provider");
        assert_eq!(provider.client_id, "cid");
        assert_eq!(provider.client_secret, "secret");
        assert_eq!(provider.token_url, "http://localhost:9999/token");
    }
}

File: nodejs/src/integration-gateway/refresh/providers.ts
import { isProdEnv } from '~/common/utils/env-utils'

/** Per-provider OAuth client credentials, sourced from the same env vars as Django. */
export interface ProviderCredentials {
    HUBSPOT_APP_CLIENT_ID: string
    HUBSPOT_APP_CLIENT_SECRET: string
    SALESFORCE_CONSUMER_KEY: string
    SALESFORCE_CONSUMER_SECRET: string
    GOOGLE_ADS_APP_CLIENT_ID: string
    GOOGLE_ADS_APP_CLIENT_SECRET: string
    GOOGLE_ANALYTICS_APP_CLIENT_ID: string
    GOOGLE_ANALYTICS_APP_CLIENT_SECRET: string
    GOOGLE_SEARCH_CONSOLE_APP_CLIENT_ID: string
    GOOGLE_SEARCH_CONSOLE_APP_CLIENT_SECRET: string
    SOCIAL_AUTH_GOOGLE_OAUTH2_KEY: string
    SOCIAL_AUTH_GOOGLE_OAUTH2_SECRET: string
    // Local/e2e mock hook only. Ignored in production (see providerFor) so a misconfigured
    // deployment can never redirect real credential traffic to an arbitrary host.
    INTEGRATION_GATEWAY_REFRESH_TOKEN_URL_OVERRIDE: string
}

/** Resolved OAuth provider config for a refreshable integration kind. */
export interface Provider {
    tokenUrl: string
    clientId: string
    clientSecret: string
}

const SALESFORCE_LOGIN_TOKEN_URL = 'https://login.salesforce.com/services/oauth2/token'

/**
 * Resolve the Salesforce token endpoint from the integration's stored `instance_url`, mirroring
 * Django's `_salesforce_instance_host`: a sandbox/My-Domain org must refresh against its own host,
 * and the prod login URL would refuse a sandbox-issued refresh_token. Validated conservatively
 * (https only, no port/userinfo, hostname under salesforce.com); anything else falls back to the
 * prod login URL.
 */
function salesforceTokenUrl(instanceUrl: unknown): string {
    if (typeof instanceUrl !== 'string' || !instanceUrl) {
        return SALESFORCE_LOGIN_TOKEN_URL
    }
    let url: URL
    try {
        url = new URL(instanceUrl)
    } catch {
        return SALESFORCE_LOGIN_TOKEN_URL
    }
    if (url.protocol !== 'https:' || url.port || url.username || url.password) {
        return SALESFORCE_LOGIN_TOKEN_URL
    }
    if (url.hostname !== 'salesforce.com' && !url.hostname.endsWith('.salesforce.com')) {
        return SALESFORCE_LOGIN_TOKEN_URL
    }
    return `https://${url.hostname}/services/oauth2/token`
}

/**
 * Resolve the OAuth provider for `kind`, or null when the kind isn't a gateway-supported OAuth2
 * provider or its client credentials aren't configured. `integrationConfig` is the row's plaintext
 * `config` (used for Salesforce's per-org instance host).
 *
 * Initial supported set is the generic `grant_type=refresh_token` providers (client id/secret in
 * the form body). Providers with non-standard refresh flows — Slack (DB-sourced creds), the
 * HTTP-Basic ones (reddit/pinterest/stripe), TikTok, Bing, Jira, Meta, and the service-account /
 * GitHub kinds — stay on the Django beat for now.
 */
export function providerFor(
    kind: string,
    config: ProviderCredentials,
    integrationConfig: Record<string, any> = {}
): Provider | null {
    let resolved: { defaultUrl: string; clientId: string; clientSecret: string } | null = null
    switch (kind) {
        case 'hubspot':
            resolved = {
                defaultUrl: 'https://api.hubapi.com/oauth/v1/token',
                clientId: config.HUBSPOT_APP_CLIENT_ID,
                clientSecret: config.HUBSPOT_APP_CLIENT_SECRET,
            }
            break
        case 'salesforce':
            resolved = {
                // Sandbox/My-Domain orgs must refresh at their own instance host (Django parity).
                defaultUrl: salesforceTokenUrl(integrationConfig.instance_url),
                clientId: config.SALESFORCE_CONSUMER_KEY,
                clientSecret: config.SALESFORCE_CONSUMER_SECRET,
            }
            break
        case 'google-ads':
            resolved = {
                defaultUrl: 'https://oauth2.googleapis.com/token',
                clientId: config.GOOGLE_ADS_APP_CLIENT_ID,
                clientSecret: config.GOOGLE_ADS_APP_CLIENT_SECRET,
            }
            break
        case 'google-analytics':
            resolved = {
                defaultUrl: 'https://oauth2.googleapis.com/token',
                clientId: config.GOOGLE_ANALYTICS_APP_CLIENT_ID,
                clientSecret: config.GOOGLE_ANALYTICS_APP_CLIENT_SECRET,
            }
            break
        case 'google-search-console':
            resolved = {
                defaultUrl: 'https://oauth2.googleapis.com/token',
                clientId: config.GOOGLE_SEARCH_CONSOLE_APP_CLIENT_ID,
                clientSecret: config.GOOGLE_SEARCH_CONSOLE_APP_CLIENT_SECRET,
            }
            break
        case 'google-sheets':
            resolved = {
                defaultUrl: 'https://oauth2.googleapis.com/token',
                clientId: config.SOCIAL_AUTH_GOOGLE_OAUTH2_KEY,
                clientSecret: config.SOCIAL_AUTH_GOOGLE_OAUTH2_SECRET,
            }
            break
        default:
            return null
    }

    if (!resolved.clientId || !resolved.clientSecret) {
        return null
    }

    // The override exists purely for local/e2e mocking; ignore it in production so a misconfigured
    // deployment fails closed instead of POSTing client secrets + refresh tokens to an arbitrary host.
    const override = config.INTEGRATION_GATEWAY_REFRESH_TOKEN_URL_OVERRIDE
    const tokenUrl = override && !isProdEnv() ? override : resolved.defaultUrl

    return {
        tokenUrl,
        clientId: resolved.clientId,
        clientSecret: resolved.clientSecret,
    }
}
